### PR TITLE
Monitor cache

### PR DESCRIPTION
**Pull request by Biboswan**
#334 
If I got it right I shall modify cache_test as well

### CODE REVIEW

**Comment by walac** [comment]
From Travis CI
```
plugins/cache/cache.go:82:30: not enough arguments in call to caching.New
	have (func(caching.Context, interface {}) (caching.Resource, error), bool, gc.ResourceTracker)
	want (caching.Constructor, bool, gc.ResourceTracker, runtime.Monitor)
plugins/cache/cache.go:83:30: not enough arguments in call to caching.New
	have (func(caching.Context, interface {}) (caching.Resource, error), bool, gc.ResourceTracker)
	want (caching.Constructor, bool, gc.ResourceTracker, runtime.Monitor)
make: *** [install] Error 2
```

File: runtime/caching/cache.go [hunk 1]
```diff
@@ -19,21 +20,27 @@ type Constructor func(ctx Context, options interface{}) (Resource, error)
...
-	shared      bool
-	entries     []*cacheEntry
-	constructor Constructor
-	tracker     gc.ResourceTracker
+	m               sync.Mutex
+	shared          bool
+	entries         []*cacheEntry
+	constructor     Constructor
+	tracker         gc.ResourceTracker
+	cachehit        runtime.Monitor
```
Comment (jonasfj) [inline runtime/caching/cache.go]: a single `monitor` property should do...

File: runtime/caching/cache.go [hunk 2]
```diff
@@ -19,21 +20,27 @@ type Constructor func(ctx Context, options interface{}) (Resource, error)
...
-func New(constructor Constructor, shared bool, tracker gc.ResourceTracker) *Cache {
+func New(constructor Constructor, shared bool, tracker gc.ResourceTracker, monitor runtime.Monitor) *Cache {
 	return &Cache{
-		constructor: constructor,
-		tracker:     tracker,
-		shared:      shared,
+		constructor:     constructor,
+		tracker:         tracker,
+		shared:          shared,
+		cachehit:        monitor.WithPrefix("cache-hit"),
```
Comment (jonasfj) [inline runtime/caching/cache.go]: Store the monitor and do `WithPrefix` when reporting... in fact you won't need it as `.Count(` takes a name

File: runtime/caching/cache.go [hunk 3]
```diff
@@ -90,6 +97,8 @@ func (c *Cache) Require(ctx Context, options interface{}) (*Handle, error) {
 
 		// Take the entry
 		debug("cache entry '%s' found in cache, refCount: %d", optionsHash, e.refCount+1)
+		c.countCachehits++
+		c.cachehit.Count("count cache-hit", c.countCachehits)
```
Comment (jonasfj) [inline runtime/caching/cache.go]: Use the `monitor` given in `New` and then call `monitor.Count('cache-hit', 1)`

If there is a cache miss, the call `monitor.Count('cache-miss', 1)`

the monitor will aggregate metrics that have the same name. Take a look at statsum if you're curious.

File: runtime/caching/cache.go [hunk 4]
```diff
@@ -111,9 +120,14 @@ func (c *Cache) Require(ctx Context, options interface{}) (*Handle, error) {
 		c.entries = append(c.entries, entry)
 
 		go entry.created.Do(func() {
-			defer entry.ctx.dispose() // ensure resources are cleanup when constructor is done
+			defer entry.ctx.dispose()
+			// ensure resources are cleanup when constructor is done
+
+			c.constructortime.Time("constructor-creation-time", func() {
```
Comment (jonasfj) [inline runtime/caching/cache.go]: Use same `monitor` as given in `New`, then call `monitor.Time('constructor-duration', func() {...`

**Comment by jonasfj** [comment]
No, I don't think they are related...

**Comment by Biboswan** [comment]
then any changes to be done?

File: plugins/cache/cache.go [hunk 5]
```diff
@@ -79,8 +79,8 @@ func (p *provider) NewPlugin(options plugins.PluginOptions) (plugins.Plugin, err
 		engine:         options.Engine,
 		environment:    options.Environment,
 		monitor:        options.Monitor,
-		sharedCache:    caching.New(constructor, false, options.Environment.GarbageCollector),
-		exclusiveCache: caching.New(constructor, false, options.Environment.GarbageCollector),
+		sharedCache:    caching.New(constructor, false, options.Environment.GarbageCollector, options.Monitor),
```
Comment (jonasfj) [inline plugins/cache/cache.go]: `.WithPrefix('shared-cache')` and simlar for exclusive...

What else are the metric names going to be?

File: runtime/caching/cache.go [hunk 6]
```diff
@@ -111,9 +116,14 @@ func (c *Cache) Require(ctx Context, options interface{}) (*Handle, error) {
 		c.entries = append(c.entries, entry)
 
 		go entry.created.Do(func() {
-			defer entry.ctx.dispose() // ensure resources are cleanup when constructor is done
+			defer entry.ctx.dispose()
+			// ensure resources are cleanup when constructor is done
```
Comment (jonasfj) [inline runtime/caching/cache.go]: This comment pertains to the line of the code that it follows? that seems unwise.

Also some grammar (which was probably there before), but `are cleaned up`

File: runtime/caching/cache.go [hunk 7]
```diff
@@ -111,9 +116,14 @@ func (c *Cache) Require(ctx Context, options interface{}) (*Handle, error) {
...
+
+			c.monitor.Time("constructor-duration", func() {
+				entry.resource, entry.err = c.constructor(entry.ctx, options)
+			})
+			//metric to give creation time of constructor
```
Comment (jonasfj) [inline runtime/caching/cache.go]: formatting... look how other comments have a space before them...

And maybe rephrase it...

File: plugins/cache/cache.go (hunk 5 above)
Comment (jonasfj) [reply plugins/cache/cache.go]: so I guess the questions is if we want them aggregated together or not...

We could get names:
```
shared-cache.cache-hit
shared-cache.cache-miss
shared-cache.constructor-duration
exclusive-cache.cache-hit
exclusive-cache.cache-miss
exclusive-cache.constructor-duration
```
or we could get:
```
cache-hit
cache-miss
constructor-duration
```
if we keep it... my question is if other place there are metrics recorded to `c.monitor` without any prefixes... I don't actually mind aggregating them, that seems decent (I think).

Comment (Biboswan) [reply plugins/cache/cache.go]: Can there be instances where you would like to see what should be shared and what should be exclusive for better performance. Otherwise better to keep them aggregated to reduce noise..., you will definitely have better idea.

Comment (jonasfj) [reply plugins/cache/cache.go]: I think reducing noise is useful... depending on how we use caches these stats could easily be somewhat irrelevant anyways.

Yes, it's cool to have them... for other caches they'll make a lot more sense..

**Comment by Biboswan** [comment]
fine its done then..

**Comment by Biboswan** [comment]
done

**Comment by jonasfj** [comment]
Thanks, this is awesome...
